### DISCUSSION TITLE
How to handle High latency networks with go-libp2p?

### DISCUSSION THREAD

**Discussion by GeorgePapageorgakis** [libp2p/go-libp2p#2731] [Q&A]
Hello, I am trying to use go libp2p to connect with other nodes in a network that has more than 3 seconds latency. I am using open WRT with Traffic Control to add latency.

However, I get the following error:

`E    * [/ip4/192.168.4.2/tcp/4001] failed to negotiate security protocol: context deadline exceeded`

I have tried some of the following:

```
	dialTimeout := 6 * time.Second // have also tried 10 minutes...
	// Create a new libp2p host/node in the network
	h, err = libp2p.New(
		libp2p.ListenAddrStrings(listenAddress, bootstrapListenAddress), identity,
		libp2p.ResourceManager(rmgr),
		libp2p.WithDialTimeout(dialTimeout), // did not help
		libp2p.SwarmOpts(swarm.WithDialRanker(swarm.NoDelayDialRanker)), // did not help
	)
```

In some cases I will just get `all dials failed` and `dial back off` errors. Also sometimes I get `all dials failed i/o timeout`.

Is this a bug, a feature by design or am I missing something?
Is there an `option` that can help p2p network play with high latency?

**Comment by MarcoPolo** [comment]
What's the RTT latency? There are a lot of hardcoded timeout assumptions right now. And some protocol negotiations take multiple round trips, so that could cause a high RTT to get multiplied and hit the timeout.

The "failed to negotiate security protocol" likely means you timed out when doing the Noise or TLS handshake. If you set `GOLOG_LOG_LEVEL=debug` you'll probably see exactly where that happens.

> Is this a bug, a feature by design or am I missing something?

For most people this is a feature. Using large timeouts opens you up to [Slowloris](https://www.cloudflare.com/learning/ddos/ddos-attack-tools/slowloris/) style attacks.

> Is there an option that can help p2p network play with high latency?

Maybe. There's a `NegotiationTimeout` in the `HostOpts` struct, but it requires building the host a bit more manually. We could have an option similar to `libp2p.WithDialTimeout` that would make this easier. 

What's your use case here?

**Reply by GeorgePapageorgakis** [reply]
Hey Marco!
RTT latency, in the actual system, is around 3 seconds. For now, I am simulating this with TC tool by setting rtt to 3 seconds.
log level is in debug. It does not show any more info. For now I do not care for Slowloris style attacks as the network does not have access to the internet.
The libp2p seems to work and connect if latency is below 700-800 ms. I am trying to increase this so that it actually play in the real scenario.
Any ideas?

**Reply by Jorropo** [reply]
I'm just curious, is there a real world connections you want to use with 3s of latency ?
In my limited experience very long RTTs also have awful packet loss and throughput.

**Reply by GeorgePapageorgakis** [reply]
Yes there are real world wireless connections with low bandwidth. This is what I am trying to do. I need to find a way to make libp2p work in this setup. For now let's ignore packet loss and keep only high latency on the table...

**Reply by MarcoPolo** [reply]
I'm not sure which timeout you're hitting. The first step to making this work for you is to figure out which timeout you need to increase. My guess is that you're hitting the [Default Negotiate timeout](https://github.com/libp2p/go-libp2p/blob/master/p2p/host/basic/basic_host.go#L49), but you could be hitting something else. Try looking through the codebase for `[dD]efault.*Timeout.*=` and adjusting those values and seeing which one needs to change for your use case to work. Then we can figure out how to make changes to the constructor options to support it (like what libp2p.WithDialTimeout does).

The core maintainers don't have time to work on this ourselves, but if a well made and well tested PR shows up I'll review and we can merge.

**Reply by GeorgePapageorgakis** [reply]
Ok I will check it. However, what other timeout options are there apart from `WithDialTimeout`. I did not find any so far...

**Reply by MarcoPolo** [reply]
I don't think we've plumbed anything else through. You have to change the library itself to figure out which timeouts need to get adjusted, and then add the plumbing to expose the new timeout options.
